src/QsCompiler/CommandLineTool/CompilationTracker.cs

```diff
@@ -0,0 +1,281 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using Newtonsoft.Json;
+
+
+namespace Microsoft.Quantum.QsCompiler.CommandLineCompiler
+{
+    /// <summary>
+    /// Provides an event tracker of the compilation process for the purpose of assessing performance. 
+    /// </summary>
+    public static class CompilationTracker
+    {
+
+        // Private classes and types.
+
+        /// <summary>
+        /// Represents a task performed by the compiler (eg. source loading, reference loading, syntax tree serialization, etc.).
+        /// </summary>
+        private class CompilationTask
+        {
+            /// <summary>
+            /// Represents the name of the parent compilation task.
+            /// </summary>
+            public readonly string ParentName;
+            /// <summary>
+            /// Represents the name of the compilation task.
+            /// </summary>
+            public readonly string Name;
+            /// <summary>
+            /// Contains the UTC datetime when the task started.
+            /// </summary>
+            public readonly DateTime UtcStart;
+            /// <summary>
+            /// Contains the UTC datetime when the task ended.
+            /// </summary>
+            public DateTime? UtcEnd;
+            /// <summary>
+            /// Contains the duration of the task in milliseconds.
+            /// </summary>
+            public long? DurationInMs;
+            /// <summary>
+            /// Stopwatch used to measure the duration of the task.
+            /// </summary>
+            private readonly Stopwatch Watch;
+
+            /// <summary>
+            /// Generates a key that uniquely identifies a task in the compilation process based on the task's name and its parent's name.
```

Comment:
> If the goal is for this to be unique, that would mean that e.g. for parsing the task name would also contain the name of the source file to parse?

Yes. In that scenario I see the parsing of all source files as one task, and each individual file parsing as a sub-task of it.

Reply:
> Also, what is the reason for wanting this to be public? I realize the class is private, but suppose at a later time it is made public, I would think that this particular method should probably remain limited to this compilation unit (e.g. because I think it's signature might change over time). Come to think of it, it seems like this method might make sense to be part of the CompilationTaskNode class instead, since that one seems to be the one that would handle the logic on how to define the tree structure for compilation tasks.

In this case it needs to be public since this method is used by the event handlers to get the key to both look for a task in the dictionary.

Even though the CompilationTaskNode is the one used to build the tree structure for compilation tasks, the key is a convenient way to store compilation tasks in a flat structure until we need to build the tree. Having said that, I think it makes sense to keep the GenerateKey method as part of the CompilationTask class.

Reply:
I changed it to internal per your suggestion.